### PR TITLE
Stream build output to the web builder during compilation

### PR DESCRIPTION
**Pull request by tomquist**
## Summary

This change enables the web builder to display the compiler's output in real-time as the build progresses, rather than only showing it after the build completes. The build workflow now publishes the compiler output as immutable segments to S3, and the web builder fetches and displays these segments as they become available.

## Key Changes

- **Build output publishing**: Added `scripts/publish-build-log.sh` to publish the compiler's output as append-only segments (`firmware/<identifier>.log.0`, `.log.1`, etc.) to S3. Since S3 doesn't support appending to objects, the log is split into immutable segments, each containing only the bytes that appeared since the last segment.

- **Job log fetching**: Added `scripts/fetch-job-log.sh` to read the build job's own log from the GitHub Actions API, clean it (removing timestamps, terminal escapes, and runner markers), and make it available for publishing. This is the only way to access the compiler's output since it runs inside a third-party action.

- **Progress watcher enhancement**: Modified `scripts/watch-build-progress.sh` to periodically call `publish-build-log.sh` alongside the existing status publishing, with a slower interval (10s vs 5s) to respect GitHub API rate limits.

- **Status document updates**: Extended `BuildStatus` to include a `logSegments` field indicating how many output segments have been published, allowing the web builder to know which segments to fetch.

- **Web builder integration**: Modified `BuildProgressDialog` to:
  - Fetch log segments as they're announced in the status document
  - Display them in a scrollable log view that auto-scrolls to the bottom unless the user has scrolled up
  - Clean segment text to remove control characters and cap segment size to 200KB for rendering performance

- **Log text cleaning**: Added `cleanLogText()` utility to sanitize compiler output by removing CRLF sequences, null bytes, and other control characters while preserving the visual output.

- **Comprehensive test coverage**: Added test suites for both `publish-build-log.sh` and `fetch-job-log.sh` covering edge cases like partial reads, upload failures, log truncation, and API unavailability.

- **Workflow permissions**: Updated GitHub Actions workflow to request `actions: read` permission so the build can read its own job log from the API.

## Implementation Details

- **Correctness guarantees**: The publishing scheme ensures every byte is uploaded exactly once, in order. The published bytes are compared against new reads to detect if the log was truncated or restarted, preventing corruption of the displayed output.

- **Best-effort design**: Log publishing is entirely optional—builds that cannot read their own log still report progress and failure, just without the output. Failed API calls don't fail the build step.

- **Rate limiting**: Log fetching uses a 10-second interval (vs 5-second for progress) to stay within GitHub API limits across multiple concurrent builds.

- **Segment caching**: The web builder downloads each segment exactly once and accumulates them, so if the API becomes unavailable mid-build, previously fetched output is retained.

https://claude.ai/code/session_01KNoarAK2zgJZG7crJUt3TH

## Summary by CodeRabbit

* **New Features**
  * Build progress now includes live compiler output while firmware builds are running.
  * Build logs load incrementally, remain available after failures, and support easier review of error details.
  * Progress updates now reflect both compiled firmware objects and available build-log output.

* **Bug Fixes**
  * Improved log publishing helps prevent missing, incomplete, or duplicated output during long-running builds.
  * Terminal formatting artifacts are now removed from displayed build logs.

* **Documentation**
  * Updated firmware storage documentation to describe build status and segmented log output.

### CODE REVIEW

File: .github/workflows/build-esphome.yml [hunk 1]
```diff
@@ -147,6 +159,31 @@ jobs:
             echo "::endgroup::"
           fi
 
+      # The watcher stopped between polls, so the last seconds of the compile -
+      # on a failed build, the error itself - are not published yet. Runs after
+      # the watcher has exited so there is only ever one writer, and before any
+      # terminal status so the page is never told about a segment that is not
+      # in the bucket yet.
+      - name: Publish the rest of the build output
```
[the remark this answers is not in the document]
Comment (tomquist) [reply .github/workflows/build-esphome.yml:167]: Confirmed and fixed in 7560a2f, though with a lock rather than a process group.

The race is real: the 60s wait has a deadline, and the `kill` after it reaches only the watcher, not a `publish-build-log.sh` it left mid-upload. Two publishers sharing `build-log.*` would both write `firmware/${IDENTIFIER}.log.${count}`, and whichever finished second would decide what that key holds — so the one with fewer bytes could replace the one with more, and the page would show a truncated log.

I took the read-and-publish under a `flock` on `${LOG_STATE_PREFIX}.lock` instead of terminating a process group. The invariant being protected is the shared state and the segment keys, not the process tree, so a lock holds however a second publisher came to exist — including the final flush racing a stray child, which is the case here. It also keeps `setsid`/PGID handling out of the workflow YAML, where `$!` is not reliably the group leader. A publisher that cannot take the lock within `LOG_LOCK_WAIT_SECONDS` (60 by default) publishes nothing and reports the count it read back, rather than racing.

Covered by a new test that holds the lock externally and asserts the publisher uploads nothing, then publishes exactly once when it is free.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/firmware/buildStatus.ts [hunk 2]
```diff
@@ -109,6 +123,38 @@ const optionalCount = (value: unknown): number | undefined =>
...
+ * Segment bodies are whatever the compiler wrote, so they arrive with stray
+ * control bytes in them. React renders the result as text either way - this is
+ * so what the user sees is what the build printed, rather than a `\r` eating
+ * the line it was on.
+ */
+export const cleanLogText = (raw: string): string => {
+  const text = raw
+    .replace(/\r\n?/g, '\n')
+    // eslint-disable-next-line no-control-regex
+    .replace(/[\u0000-\u0008\u000b-\u001f\u007f]/g, '');
```
[the remark this answers is not in the document]
Comment (tomquist) [reply src/firmware/buildStatus.ts:160]: Confirmed and fixed in 7560a2f. Reproduced first — `printf 'a\033[31mred\033[0m b\033[1:2mfancy\033[0m'` through `fetch-job-log.sh` left `^[[1:2m` in the output, exactly as described. `[0-9;?]` omits `:` (and `<` `=`), and `[a-zA-Z]` is narrower than the `@-~` final-byte range.

Fixed in both places rather than only the browser: the shell pattern is now the whole CSI form (`[0-?]*[ -/]*[@-~]`), and `cleanLogText` does the same as the last line of defence, since it is the only point that sees what a segment actually contains. Test expectation updated as you noted, plus a new test in `fetch-job-log.test.mjs` covering the sub-parameter form alongside the plain and truecolor ones.

Replaying a real 533 KB job log now publishes with zero escape bytes left in it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/components/BuildProgressDialog.tsx [hunk 3]
```diff
@@ -92,6 +93,52 @@ const compiledFiles = (status: BuildStatus | null): string | null => {
...
 
+/**
+ * The tail of the build output, pinned to its last line.
+ *
+ * Anchored only while it is already at the bottom, so a user who scrolled up to
+ * read an error is not yanked back down by the next update - which arrives
+ * every few seconds for the length of the compile.
+ */
+const BuildLog: React.FC<{ log: string }> = ({ log }) => {
+  const box = useRef<HTMLElement | null>(null);
```
[the remark this answers is not in the document]
Comment (tomquist) [reply src/components/BuildProgressDialog.tsx]: Applied in 7560a2f, though the stated failure doesn't occur: `npm run typecheck` passes with `useRef<HTMLElement | null>(null)` both locally and in CI (the Typecheck step was green on 3635e44). MUI's `OverridableComponent` accepts the wider ref type here.

Taking it anyway as a precision nit — `HTMLPreElement` is what `Box component="pre"` infers, so it's the more accurate thing to write, and it costs nothing. Typecheck still passes.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/publish-build-log.test.mjs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+  space.write('Compiling app\n');
+
+  const holder = spawn(
+    'flock',
+    [path.join(space.dir, 'state.lock'), 'sleep', '5'],
+    { stdio: 'ignore' }
+  );
+  try {
+    // Give flock a moment to actually take it before racing it.
+    execFileSync('bash', ['-c', 'sleep 0.5']);
```
[the remark this answers is not in the document]
Comment (tomquist) [reply scripts/publish-build-log.test.mjs]: Right, and the mechanism is worse than the one described. `holder.kill()` doesn't just fail to *wait* — it doesn't *release*: `flock` passes the descriptor it holds to the command it runs, so `sleep` inherits the lock and keeps holding it after flock is gone. Demonstrated directly with a non-blocking probe:

```
locked while holder runs: true
still locked after holder.kill(): true
```

So the second `publish()` wasn't racing a shutdown, it was contending a lock that was never coming back — it only passed because the publisher falls back to publishing when it can't take the lock within the default 60s wait, which is not what that assertion was meant to prove.

Fixed in f166841: the holder is spawned `detached` and its process *group* is signalled, the test waits on a non-blocking `flock -n` probe to confirm the lock was actually taken before contending it, and waits for release before publishing again. No sleeps anywhere. It's also skipped outright where `flock` isn't installed — the same case in which the script publishes without a lock at all, so there'd be nothing to assert.

Five consecutive runs clean.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (tomquist) [reply scripts/publish-build-log.test.mjs]: Correcting one thing I said above: there is no "falls back to publishing" — on lock timeout the publisher reports the count and uploads nothing, which is the safety property the fix relies on.

What actually happened in the old test is that the orphaned `sleep 5` held the lock for its remaining ~4.5s, and the second `publish()` simply blocked on flock until it expired and then acquired it legitimately. So the test passed by waiting out a timer rather than by observing a release — slow, and dependent on `sleep 5` being shorter than the 60s default wait. Change either number and it breaks. The fix and everything else above stand.

---
_Generated by [Claude Code](https://claude.ai/code)_
